### PR TITLE
[OFFAPPS-1023] No validation for "Start" and "Date" fields if Created/Updated not selected

### PR DESCRIPTION
**Pull request by yu-tian113**
[OFFAPPS-1023] No validation for "Start" and "Date" fields if Created/Updated not selected

## Description
"Date range", "Start" and "End" fields don't have validation on values and logics. @m-lai has created a new design to improve the UX. 

## Tasks
- [x] Change "Start" and "End" fields to be type `date`, adding validations for browsers not support Form constraint API
- [x] Update as per new UX design(both looking and feature)
- [x] Write tests

## Implementation notes
[closest](https://www.npmjs.com/package/closest) is used to handle event delegation

## References
[JIRA ticket](https://zendesk.atlassian.net/browse/OFFAPPS-1023)
[Invision design](https://zendesk.invisionapp.com/share/8DJQOTNNT9M#/screens/299998016_sidebar-Search-App-Specs)

#### Date range none
None
#### Safari
Safari
#### Chrome
Chrome

## CCs
@m-lai 
@zendesk/apps-migration

## Risks
* [medium] Regression testing and new feature testing

### CODE REVIEW

**Comment by m-lai** [comment]
Hey @yu-tian113 , if "None" is selected for the "Date Range" dropdown, then we should collapse the "Start" and "End" fields. But otherwise the validation looks correct if a legit value is selected under the "Date Range" dropdown

**Comment by yu-tian113** [comment]
@m-lai ah, we do. Sorry I was being lazy using the screenshot for translation team, where I put two strings on the same pic.

File: src/templates/results.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const getTicketMarkup = (o) => {
     `
     <tr class="c-table__row">
       <td class="c-table__row__cell u-position-relative">
-        <a href class="ticket-link" data-id="${o.id}"><b data-id="${o.id}">#${o.id}</b> ${escape(o.subject)}</a>
```
Comment (hoangtrvu) [inline src/templates/results.js]: Is `data-id` in `<b>` no longer used?

File: src/templates/search.js [hunk 2]
```diff
@@ -63,23 +63,24 @@ export default function (args) {
...
+          <div class="row date-range--fields">
             <div class="col">
               <label class="c-txt__label c-txt__label--sm" for="from">${I18n.t('search.filter.date_start')}</label>
-              <input type="text" id="from" class="c-txt__input c-txt__input--sm" placeholder="YYYY-MM-DD">
+              <input type="date" id="from" class="c-txt__input c-txt__input--sm" placeholder="YYYY-MM-DD" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}">
             </div>
             <div class="col">
               <label class="c-txt__label c-txt__label--sm" for="to">${I18n.t('search.filter.date_end')}</label>
-              <input type="text" id="to" class="c-txt__input c-txt__input--sm" placeholder="YYYY-MM-DD">
+              <input type="date" id="to" class="c-txt__input c-txt__input--sm" placeholder="YYYY-MM-DD" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}">
```
Comment (hoangtrvu) [inline src/templates/search.js]: I thought date input is not supported on IE <12 https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date#Browser_compatibility?

Comment (sostopher) [reply src/templates/search.js]: It isn't, that's why there's a fallback to input field and validation done on that.

File: src/javascript/modules/search.js [hunk 3]
```diff
@@ -158,16 +166,19 @@ class Search {
...
         {
           assignees: await this._getAssignees().catch(this._handleRequestFail.bind(this, '.loader'))
         }
       )
       await this._render('#assignee', getAssigneesTemplate)
+      this._hideLoadingButton()
     }
     Object.assign(this._states, {showAdvancedOptions: event.target.checked})
-    this._appContainer.querySelector('.advanced-options-wrapper').classList.toggle('u-display-block')
+    if(!event.target.checked) this._resetDateFields()
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/modules/search.js]: fixed!

File: src/templates/results.js (hunk 1 above)
Comment (yu-tian113) [reply src/templates/results.js]: correct, don't need anymore. [closest](https://github.com/ForbesLindesay/closest) is used to find parent that should received the event delegation

File: src/templates/search.js (hunk 2 above)
Comment (yu-tian113) [reply src/templates/search.js]: yes, IE/Safari doesn't support date which fall back to `type="text"` and we use regex pattern to validate the value

**Comment by fvzzy** [comment]
Do we need to use https://www.npmjs.com/package/closest? It doesn't seem very actively maintained, and seems like something we could just implement ourselves https://github.com/ForbesLindesay/closest/blob/master/index.js

File: src/main.css [hunk 4]
```diff
@@ -69,6 +73,27 @@ div.c-txt__input {
...
+    .date-range--fields {
+      display: flex;
+    }
+  }
+  &.show-error {
+    .date-range--error {
+      display: block;
+    }
+  }
+  .col{
```
Comment (keitheous) [inline src/main.css]: I think we missed a space. 

`.col {`

Comment (yu-tian113) [reply src/main.css]: good catch, fixed

**Comment by yu-tian113** [comment]
@fvzzy, yes, I'm happy to spend some time to build our own.
Will need to implement the dependency [matches-selector](https://github.com/ForbesLindesay/matches-selector) as well for cross browser support.

File: spec/search.spec.js [hunk 5]
```diff
@@ -100,36 +100,22 @@ describe('Search App', () => {
...
-      app._keywordField.dispatchEvent(new KeyboardEvent('keydown', { 'which': 13 }))
-      expect(doTheSearchSpy).toHaveBeenCalledTimes(1)
-      app._keywordField.dispatchEvent(new KeyboardEvent('keydown', { 'which': 32 }))
+      app.$searchForm.dispatchEvent(new Event('submit'))
       expect(doTheSearchSpy).toHaveBeenCalledTimes(1)
     })
 
-    it('should trigger search when search button is clicked', () => {
-      app._client.request.mockReturnValue(Promise.resolve(RESULTS_1))
-      document.querySelector('#search-submit svg').dispatchEvent(
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply spec/search.spec.js]: The test is removed, we are watching form submit event now.
But yes, it should work.

File: src/javascript/modules/search.js [hunk 6]
```diff
@@ -138,18 +152,12 @@ class Search {
...
-   * @param {Event} event
-   */
-  _handleKeydown (event) {
-    if (event.which === 13) this._doTheSearch(event)
+    let closestTicketLink
+    if (closest(target, '.suggestion', true)) this._handleSuggestionClick(event)
+    else if (target.dataset.index && closest(target, '.page-link', true)) this._doTheSearch(event, target.dataset.index)
+    // add an extra property dispatchTarget to click event to store the target element,
+    // click event may be triggered by child elements of our target element
+    else if (closestTicketLink = closest(target, '.ticket-link', true)) this._handleResultLinkClick(Object.assign(event, {dispatchTarget: closestTicketLink}))
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/modules/search.js]: updated!

File: spec/search.spec.js [hunk 7]
```diff
@@ -54,7 +54,7 @@ describe('Search App', () => {
       app = new Search(CLIENT, APPDATA_WITH_CF, CONFIG)
       doTheSearchSpy = jest.spyOn(app, '_doTheSearch')
       app._initializePromise.then(() => {
-        app._keywordField.value = 'TestKeyword'
+        app.$keywordField.value = 'TestKeyword'
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply spec/search.spec.js]: Yes, we have it covered in the code refactor.

**Pull request merged**